### ISSUE TITLE
📎 Vue Semantic Model

### ISSUE THREAD

**Issue by dyc3** [biomejs/biome#10211]
## Problem

It's become clear that the `EmbeddedValueReferences` and `EmbeddedExportedBindings` services aren't quite the way to go to deal with cross language lint rules.

### Correctness and Maintainability

The shared services work across multiple template languages (Vue, Svelte, Astro). On one hand, this theoretically makes it easier to support all template languages, and there's a lot of overlapping concepts. However, they have differences that can introduce false negatives.

For example, when using the Vue Options API, members of `props`, `components`, `data`, `methods`, and `watch` fields become accessible bindings from the template section. But they are never defined like a normal variable. We currently special case only `props` https://github.com/biomejs/biome/pull/9285, but this is a rather hacky solution, which was unable to use our existing vue helper utilities.

Another quirk of our current architecture is how `v-for` is handled. Bindings created by a v-for directive are local to the scope of that element. They are available to v-bind directives like `:key`, [but _not_ to `v-if` directives.](https://vuejs.org/guide/essentials/list.html#v-for-with-v-if) However, Biome currently treats those bindings as globally available in the component, which results in false negatives.

```vue
{{ i }} // should be considered undeclared
<div v-for="i in 10">{{i}}</div>
```

```vue
// `i` in the v-if should be considered undeclared
<div v-for="i in 10" v-if="i !== 5">{{i}}</div>
```

I suppose we could work in these concepts into the existing services, but I see that code becoming very complicated very fast. Having separate component semantic models for each template language would let us handle each language in isolation, and reduce the risk of a fix for one language causing regressions in another. 

With the current generic approach, we also loose information about created bindings. It would be valuable to know where a binding came from for certain rules. For example, it's OK to write to something declared in `data`, but not to a one declared in `computed`. But these concepts don't matter or apply in different ways to Svelte/Astro.

### Performance

Currently, Vue rules in the js analyzer use a series of lazy helpers to extract information and make assertions. (`crates/biome_js_analyze/src/frameworks/vue/vue_component.rs`) It usually means that all the rules start with this query:

```
/// VueComponentQuery is a query type that can be used to find Vue components.
/// It can match any potential Vue component.
pub type VueComponentQuery = Semantic<AnyPotentialVueComponent>;
```

This means that every rule has to repeat work:
1. to assert: is this a vue component?
2. to assert: what kind of vue component is it?
3. and to lookup/recollect lists of declarations

This could all be done ahead of time to reduce the repeated work.

## Solution

We want to have a Vue specific semantic model that can hold information about vue components as a whole. So far we need to track:
- bindings exported from `<script>` section (for vue, this varies a lot depending on how the component is defined)
- bindings created inside the template section (like `v-for` and slots) and their scopes
- references to bindings
- usages of emitted events (to be able to implement `noUnusedEmitDeclarations`)

There's probably other things. That's what I created this issue to discuss.

**Comment by ematipico** [comment]
> services aren't quite the way to go to deal with cross language lint rules.

It's not very clear why though. It's okay to shift to another solution , but if you don't explain the reasons or limitations, there can't be a discussion.

The good thing about these services, is that they are language agnostic, they are portable and they can used anywhere. In fact they work for all languages. Let's use it as a starting point.

In the second section of the description, you suggested what we already have today, which is a given. 

So have a counter point to discuss: what about other languages? And about the rules? Why Vue and not svelte too?

**Comment by dyc3** [comment]
I'll edit it to elaborate a bit more

**Comment by ematipico** [comment]
I see the concern is real, so I agree we should have proper services that support the specific languages. However, there's a concrete limitation that we need to take into account. These services need to span multiple languages to be constructed **and** shared, which means they need to be language-agnostic. Worst-case scenario, we would need to create an HIR to represent them, if it comes to it. That's what we do with our type-inference.
